File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static boolean[] visited;
    public int solution(int n, int[][] computers) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int i = 0; i < n; i++) graph.put(i, new ArrayList<>());
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i != j && computers[i][j] == 1)
                    graph.get(i).add(j);
            }
        }
        
        visited = new boolean[n];
        int answer = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                // bfs(n, computers, i, graph);
                dfs(n, computers, i, graph);
                answer++;
            }
        }
        return answer;
    }
    
    private void bfs(int n, int[][] computers, int start, Map<Integer, List<Integer>> graph) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        visited[start] = true;
        while (!q.isEmpty()) {
            int node = q.poll();
            for (int nextNode : graph.get(node)) {
                if (!visited[nextNode]) {
                    q.add(nextNode);
                    visited[nextNode] = true;
                }
            }
        }
    }
    
    private void dfs(int n, int[][] computers, int start,  Map<Integer, List<Integer>> graph) {
        Stack<Integer> s = new Stack<>();
        s.push(start);
        visited[start] = true;
        while (!s.isEmpty()) {
            int node = s.pop();
            for (int nextNode : graph.get(node)) {
                if (!visited[nextNode]) {
                    s.push(nextNode);
                    visited[nextNode] = true;
                }
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import defaultdict, deque
import heapq


def solution(n, computers):
    visited = set()
    graph = defaultdict(list)
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] == 1:
                graph[i].append(j)
                graph[j].append(i)
    
    def dfs(start):
        stack = []
        stack.append(start)
        visited.add(start)
        while stack:
            node = stack.pop()
            for next_node in graph[node]:
                if next_node not in visited:
                    visited.add(next_node)
                    stack.append(next_node)
            
    def bfs(start):
        q = deque()
        q.append(start)
        visited.add(start)
        while q:
            node = q.popleft()
            for next_node in graph[node]:
                if next_node not in visited:
                    visited.add(next_node)
                    q.append(next_node)
        
    answer = 0
    for i in range(n):
        if i not in visited:
            # dfs(i)
            bfs(i)
            answer += 1
    
    return answer